File: python/python/soma_next/worker.py
"""The generic worker: `pip install soma-next` and nothing else.

    python -m soma_next.worker --listen 127.0.0.1:7000 [--store /scratch/soma]

An independent process, in the background, on whatever machine. It starts
**empty** — it does not know what `tokenize` is — and waits for someone to
connect and send it what to build its catalog from. There is no way of giving it
a catalog by hand, and that is on purpose: **a worker receives the plan and
resolves the implementations**.

## The strategies, and how they differ

| artifact kind | what arrives | what the worker supplies |
|---|---|---|
| `project` *(default)* | names, versions and state | **the code**, from its clone |
| `pickle` | the code and the state | nothing |

`project` is the one you want when the worker runs in a clone of the project: it
sends tens of bytes per node, couples no interpreters, and **checks the version**
— if your clone is half-updated, it finds out before executing. `pickle` removes
all friction when the worker does not have your code: a bare node, a
`pip install`, and that is it. A worker accepts both, by `kind`.

With `project`, if the class here is not the version the graph was written
against, `--strict` *(default)* stops and says so with both versions in front of
you, and `--lucky` executes whatever it has and **reports it on `stderr`** —
running a different version silently is what gets discovered three days later.

## The port belongs on a network you trust

A worker executes what it is sent, and that is the whole point of it — but say
it plainly: `pickle` artifacts are opened with `cloudpickle.loads`, and `project`
ones resolve classes out of this clone, so **whoever reaches this port runs code
here**, as their user. There is no authentication and there is not going to be
one: this is `srun`'s and `ssh`'s job, not a framework's.

Bind it to `127.0.0.1` and reach it through a tunnel, or to a private interface
inside a cluster. `0.0.0.0` on a machine with a public address is handing out a
shell.

## What this module does NOT solve

The **environment**. `cloudpickle` moves your objects, it does not move `torch`.
If the dependencies your nodes import are not installed here, the `loads` fails
with an ordinary `ModuleNotFoundError`. Installing them belongs to whoever stands
the worker up, and putting it in here cost the original soma 420 lines of
environment manager and a hot `pip install`.

Which is a **recipe and not a mechanism**, and it fits in one file::

    #!/usr/bin/env -S uv run --script
    # /// script
    # requires-python = "==3.13.*"
    # dependencies = ["soma-next[remote]", "torch==2.10.0"]
    # ///
    from soma_next import worker
    worker.listen("0.0.0.0:7000", store="/scratch/soma")

`uv lock --script` leaves the resolution beside it, so the machine that claims a
trial from a shared folder gets the environment the one that wrote it had —
without a shared env on NFS and without a `module load`. Nothing in here knows
what `uv` is, and that is the point: it is the same file the worker already was.

## `stdout` is the wire

The protocol's messages go over this process's standard output, so at startup
`sys.stdout` is redirected to `sys.stderr`: a stray `print()` in one of your
nodes would otherwise break them. A library that writes to file descriptor 1
directly would still break it; that is rare and we do not paper over it.
"""

from __future__ import annotations

import sys

from soma_next import _manifest
from soma_next._soma_next import listen_provisioned as _listen_provisioned
from soma_next._soma_next import serve_provisioned as _serve_provisioned

__all__ = [
    "Pickles",
    "Project",
    "Strategies",
    "listen",
    "runtime",
    "serve_provisioned",
]


def runtime():
    """How this process identifies itself to the other side, so a mismatch is
    refused on connect instead of surfacing inside a `loads`.

    `soma_next`'s version goes in here, and that is why it does not go into each
    class's fingerprint.
    """
    import cloudpickle

    from soma_next import __version__

    v = sys.version_info
    return (
        f"cpython-{v.major}.{v.minor}"
        f"/cloudpickle-{cloudpickle.__version__}"
        f"/soma-next-{__version__}"
    )


class Pickles:
    """Turns `pickle` artifacts into nodes. Python's `Provision`:
    `accepts(client, kind)` returns `None` or **how this worker identifies
    itself**, and `provide(kind, blob)` returns the nodes as a dict."""

    def accepts(self, client, kind):
        del kind  # `Pickles` only opens one, and checks it in `catalog`
        mine = runtime()
        return None if client == mine else mine

    def provide(self, kind, blob):
        if kind != "pickle":
            raise ValueError(
                f"this worker only knows how to open `pickle` artifacts, and a `{kind}` arrived"
            )
        import cloudpickle

        try:
            sent = cloudpickle.loads(blob)
        except ModuleNotFoundError as e:
            # The likeliest failure: the nodes came from a client module that
            # does not exist here, so cloudpickle stored a reference.
            raise ModuleNotFoundError(
                f"this worker does not have the module `{e.name}`, so it cannot open "
                f"what it was sent. Either install it here, or make it travel inside "
                f'the artifact: Worker.generic(..., send=["{(e.name or "").split(".")[0]}"])'
            ) from e

        return _nodes(sent, "pickle")


class Project:
    """Opens `project` artifacts: resolves the classes against **this** clone.
    No code comes over the wire, so neither `cloudpickle` nor matching
    interpreters are needed."""

    def __init__(self, strict=True):
        self.strict = strict

    def accepts(self, client, kind):
        del client, kind
        # Nothing to refuse: no code arrives, so the interpreter does not
        # matter. The version is checked on opening, class by class.
        return None

    def provide(self, kind, blob):
        if kind != _manifest.KIND:
            raise ValueError(f"this does not open `{kind}` artifacts")
        return _nodes(_manifest.unpack(blob, strict=self.strict), _manifest.KIND)


class Strategies:
    """Several ways of building the catalog, asked by `kind` and not tried in a
    chain to see which sticks."""

    def __init__(self, **by_kind):
        self.by_kind = by_kind

    def _who(self, kind):
        if (which := self.by_kind.get(kind)) is None:
            raise ValueError(
                f"this worker cannot open `{kind}` artifacts; it can open: "
                + ", ".join(sorted(self.by_kind))
            )
        return which

    def accepts(self, client, kind):
        return self._who(kind).accepts(client, kind)

    def provide(self, kind, blob):
        return self._who(kind).provide(kind, blob)


def _nodes(sent, kind):
    """What an artifact has to unpack to: a dict of id → node."""
    if not isinstance(sent, dict):
        raise TypeError(
            f"a `{kind}` artifact is a dict of id → node, and a "
            f"`{type(sent).__name__}` arrived"
        )
    return sent


def default(strict=True):
    """What a worker knows how to open if you do not say otherwise."""
    return Strategies(**{_manifest.KIND: Project(strict), "pickle": Pickles()})


def serve_provisioned(provision=None, store=None):
    """Serves slices with the catalog the client sends, until the client closes.
    Without an argument it opens both kinds."""
    _hush()
    return _serve_provisioned(
        default() if provision is None else provision, store=store
    )


def _hush():
    """Leaves `stdout` free for the protocol and sends what is printed to `stderr`."""
    sys.stdout = sys.stderr


def listen(addr, provision=None, opened=None, store=None):
    """Stands on `addr` and serves whoever connects. It does not return.

    `provision` says what the implementations are resolved with; by default,
    `project` and `pickle`. `stdout` is not touched — here the wire is the
    socket — and `opened` is called once with the real address, so port `0` can
    be asked for.

    `store` is a directory, and it answers two questions that stay two: an
    artifact it already has is **not sent again**, and a node whose answer is
    already there is **not run again**. Shared between workers — a mount, a
    network folder — the second one to be stood up starts warm.
    """
    return _listen_provisioned(
        addr, default() if provision is None else provision, opened, store=store
    )


def main(argv=None):
    """`python -m soma_next.worker [--listen HOST:PORT] [--store DIR] [--lucky]`.

    Without `--listen` it talks over standard input, which is for testing.
    `--lucky` executes even if its code is not the version the graph was written
    against; by default it stops. `--store` is a directory to keep things in.
    """
    argv = list(sys.argv[1:] if argv is None else argv)
    _read_the_codecs()
    which = default(strict="--lucky" not in argv)
    store = _after("--store", argv)
    if "--listen" in argv:
        return listen(
            _after("--listen", argv, needed=True),
            provision=which,
            opened=lambda where: print(f"listening on {where}", flush=True),
            store=store,
        )
    return serve_provisioned(which, store=store)


def _read_the_codecs():
    """Registers what this worker knows how to read, **before** it serves.

    A codec has to be there before a value arrives and not be discovered while
    one is being unpacked. Discovered, the first import of a large native
    extension happens inside a request thread while the GIL is being handed
    back and forth, and with torch that is not an error but **heap corruption**:
    `free(): chunks in smallbin corrupted`, a worker that dies with no
    traceback, and a client told only that it closed without answering.

    It hid for a long time because the first job a worker got was usually a
    small one, which imported torch harmlessly; it surfaced the day a second
    torch worker was stood up whose very first job was a training run.

    This is not the worker learning about your project. `soma_next.torch` is
    soma-next's own, and a worker that can carry a tensor is one that has read
    the codec for it. One that has no torch carries no tensors, and that is a
    worker too.
    """
    try:
        import soma_next.torch  # noqa: F401
    except ImportError:
        pass


def _after(flag, argv, needed=False):
    """What comes after a flag, or `None` if the flag is not there."""
    if flag not in argv:
        return None
    after = argv.index(flag) + 1
    if after >= len(argv):
        raise SystemExit(f"`{flag}` needs a value: {flag} 127.0.0.1:7000")
    return argv[after]


if __name__ == "__main__":
    main()

File: transport/src/worker.rs
//! A process that gets sent work. This side.
//!
//! The first real implementation of [`Transport`], and it does what the GIL will
//! not allow with threads: two Python nodes in the same wave interleave but do
//! not overlap, and in two processes they do.
//!
//! **The session opens by itself, and once.** The greeting is sent not when the
//! process starts but **before the first job**: a worker that stands up and
//! receives nothing should have done nothing. After that, [`dispatch`] does not
//! greet again while the conversation is alive.
//!
//! [`dispatch`]: Transport::dispatch
//!
//! **A worker serves one at a time.** [`Transport`] is `Sync`, so two branches
//! of a wave can call `dispatch` at once — and a pipe does not fit two
//! conversations. The `Mutex` queues them, which is correct and not a
//! limitation: a worker is **one** process. If you want two at once, stand up
//! two workers and place them on two hosts.
//!
//! # Two ends, and the one that matters is the second
//!
//! ```ignore
//! Worker::spawn(Command::new("./my-worker"))   // a child, over pipes
//! Worker::connect("node3:7000")                // one that was already standing
//! ```
//!
//! The first is convenient for testing and **does not satisfy the use case**: as
//! long as the client starts the process, there is no independent worker worth
//! the name. The second is the real form. That the conversation never finds out
//! which it is falls out of [`frame`](crate::frame) working over
//! `impl Read`/`impl Write`.
//!
//! [`Worker::spawn`] takes a ready-made [`Command`] rather than a path because
//! this library does not know what your binary is called, nor what environment
//! it needs, nor whether it goes inside an `srun`.

use crate::codec::{self, Codec};
use crate::frame;
use crate::{Answer, Artifact, Request};
use soma_next_core::{Cargo, Outcome, Plan, Transport, TransportError, Watcher};
use std::io::{self, BufReader};
use std::net::{TcpStream, ToSocketAddrs};
use std::process::{Child, ChildStdin, ChildStdout, Command, Stdio};
use std::sync::{Arc, Mutex};

/// A process that executes the slices it is sent.
pub struct Worker {
    /// The process, its two ends and how far the conversation has got, held at
    /// once so two threads do not cross halfway through a message.
    open: Mutex<Open>,
    /// What to provision it with and how this client identifies itself; `None`
    /// is a worker that brings its own catalog. Behind a lock because it is set
    /// **after** opening: which nodes go here is known at run time.
    carries: Mutex<Option<(Artifact, String)>>,
    /// Who writes down what would not otherwise cross. `None` is the whole of
    /// Rust: there, an opaque carries something nobody has said how to write.
    ///
    /// Owned and not lent, unlike [`Serving`](crate::Serving)'s: this type has
    /// no lifetime and is held inside an `Arc` by whoever executes.
    codec: Option<Arc<dyn Codec>>,
}

struct Open {
    link: Link,
    /// The session opens once, not once per job.
    greeted: bool,
}

/// How the worker is spoken to. Both variants do the same thing; what changes
/// is who started the process and who ends it.
enum Link {
    /// A child process, over its pipes.
    Child {
        child: Child,
        /// `Option` so it can be closed on its own: that is the child's signal
        /// that there is no more work.
        to: Option<ChildStdin>,
        from: BufReader<ChildStdout>,
    },
    /// A worker that was already standing. Two handles on the same socket, so
    /// it can write and read without fighting over a borrow.
    Socket {
        to: TcpStream,
        from: BufReader<TcpStream>,
    },
}

impl Link {
    fn send(&mut self, payload: &[u8]) -> io::Result<()> {
        match self {
            Self::Child { to, .. } => match to {
                Some(to) => frame::send(to, payload),
                None => Err(io::Error::new(
                    io::ErrorKind::BrokenPipe,
                    "this worker is already closed",
                )),
            },
            Self::Socket { to, .. } => frame::send(to, payload),
        }
    }

    fn recv(&mut self) -> io::Result<Option<Vec<u8>>> {
        match self {
            Self::Child { from, .. } => frame::recv(from),
            Self::Socket { from, .. } => frame::recv(from),
        }
    }
}

impl Worker {
    /// Starts the process and keeps its pipes. The worker there has to bring
    /// its own catalog; for an empty one, see [`Worker::carrying`].
    ///
    /// `stderr` is left **inherited**: its `stdout` is the wire.
    pub fn spawn(mut command: Command) -> io::Result<Self> {
        let mut child = command
            .stdin(Stdio::piped())
            .stdout(Stdio::piped())
            .spawn()?;
        let to = child.stdin.take().expect("just asked for it with `piped`");
        let from = child.stdout.take().expect("just asked for it with `piped`");
        Ok(Self::over(Link::Child {
            child,
            to: Some(to),
            from: BufReader::new(from),
        }))
    }

    /// Connects to a worker that was already running — the form that satisfies
    /// the use case. On the other side there is [`Serving::listen`](crate::Serving::listen).
    pub fn connect(addr: impl ToSocketAddrs) -> io::Result<Self> {
        let to = TcpStream::connect(addr)?;
        // No `Nagle`: the messages are small and go ping-pong.
        to.set_nodelay(true)?;
        let from = BufReader::new(to.try_clone()?);
        Ok(Self::over(Link::Socket { to, from }))
    }

    /// The same worker, carrying what to provision it with if it started empty.
    /// `runtime` — `cpython-3.13/cloudpickle-3.1` — is what the far side's
    /// [`Provision`](crate::Provision) can reject at greeting time.
    pub fn carrying(self, artifact: Artifact, runtime: impl Into<String>) -> Self {
        let _ = self.offering(artifact, runtime);
        self
    }

    /// The same on an already-built worker, for whoever decides at run time
    /// which nodes go to this host. Setting the same artifact twice does
    /// nothing; changing an open session's fails.
    pub fn offering(
        &self,
        artifact: Artifact,
        runtime: impl Into<String>,
    ) -> Result<(), TransportError> {
        let mut carries = match self.carries.lock() {
            Ok(carries) => carries,
            Err(poisoned) => poisoned.into_inner(),
        };
        if let Some((already, _)) = carries.as_ref() {
            if already.id == artifact.id {
                return Ok(());
            }
            if self.greeted() {
                return Err(TransportError::new(format!(
                    "this worker already opened a session with `{}` and its catalog \
                     cannot be changed to `{}` without reconnecting",
                    already.id, artifact.id
                )));
            }
        }
        *carries = Some((artifact, runtime.into()));
        Ok(())
    }

    /// The same worker, with somebody who knows how to write down what an
    /// opaque carries.
    ///
    /// Without one, a value that only exists in this process is refused at
    /// encoding time, as it always was. With one, it crosses as bytes and the
    /// refusal is left for what nobody registered a codec for.
    pub fn packing(mut self, codec: Arc<dyn Codec>) -> Self {
        self.codec = Some(codec);
        self
    }

    fn greeted(&self) -> bool {
        match self.open.lock() {
            Ok(open) => open.greeted,
            Err(poisoned) => poisoned.into_inner().greeted,
        }
    }

    fn over(link: Link) -> Self {
        Self {
            open: Mutex::new(Open {
                link,
                greeted: false,
            }),
            carries: Mutex::new(None),
            codec: None,
        }
    }
}

impl Open {
    /// Sends a message and waits for the one that answers it, handing whatever
    /// the far side says on the way to `seen`.
    ///
    /// It reads **until an answer is terminal**, which is the one change the
    /// whole live half of this needed: [`Answer::Saw`] is not an answer to
    /// anything, it is the worker talking while it works, and the loop is what
    /// turns a blocked read into a stream.
    ///
    /// A fact is passed on exactly as it was emitted. Attributing it to a host
    /// is the engine's job, not this one's: here the host is an address, and the
    /// name the graph gave it is not known.
    fn say(
        &mut self,
        request: &Request,
        seen: Option<&dyn Watcher>,
    ) -> Result<Answer, TransportError> {
        let payload = request
            .to_bytes()
            .map_err(|e| TransportError::new(e.to_string()))?;

        self.link.send(&payload).map_err(|e| broke(&e))?;
        loop {
            let answer = self
                .link
                .recv()
                .map_err(|e| broke(&e))?
                .ok_or_else(|| TransportError::new("the worker closed without answering"))?;
            match Answer::from_bytes(&answer).map_err(|e| TransportError::new(e.to_string()))? {
                // Not an answer: keep waiting for one. A client that is not
                // watching still has to read these off the socket — dropping
                // them is what it means not to watch, and leaving them there
                // would desynchronise the conversation.
                Answer::Saw(fact) => {
                    if let Some(seen) = seen {
                        seen.saw(&fact);
                    }
                }
                terminal => return Ok(terminal),
            }
        }
    }

    /// Opens the session, if it was not open: the artifact's **name** is
    /// announced and the bytes only sent if the worker asks for them.
    fn greet(
        &mut self,
        runtime: &str,
        artifact: Option<&Artifact>,
        seen: Option<&dyn Watcher>,
    ) -> Result<(), TransportError> {
        if self.greeted {
            return Ok(());
        }
        let hello = Request::Hello {
            runtime: runtime.to_string(),
            offering: artifact.map(Artifact::label),
        };
        match self.say(&hello, seen)? {
            Answer::Ready => {}
            Answer::Send => {
                let artifact = artifact.ok_or_else(|| {
                    TransportError::new(
                        "the worker asked for an artifact and this client brings none",
                    )
                })?;
                let sending = Request::Provision {
                    bytes: artifact.bytes.clone(),
                };
                match self.say(&sending, seen)? {
                    Answer::Ready => {}
                    other => return Err(unexpected("provisioning", &other)),
                }
            }
            other => return Err(unexpected("greeting", &other)),
        }
        self.greeted = true;
        Ok(())
    }
}

impl Transport for Worker {
    fn dispatch(
        &self,
        plan: &Plan,
        cargo: &Cargo<'_>,
        seen: Option<&dyn Watcher>,
    ) -> Result<Outcome, TransportError> {
        let carries = self
            .carries
            .lock()
            .map_err(|_| TransportError::new("this worker was poisoned by an earlier panic"))?;
        let mut open = self
            .open
            .lock()
            .map_err(|_| TransportError::new("this worker was poisoned by an earlier panic"))?;

        match carries.as_ref() {
            Some((artifact, runtime)) => open.greet(runtime, Some(artifact), seen)?,
            None => open.greet("rust", None, seen)?,
        }
        drop(carries);

        // Written down before the message is built, so the refusal in
        // `Request::to_bytes` is untouched and still guards: by the time it
        // looks, whatever had a codec is already bytes. Nothing to write down
        // means nothing is copied.
        let (input, known) = match self.codec.as_deref() {
            None => (cargo.input.clone(), cargo.known.to_vec()),
            Some(codec) => (
                codec.packed(cargo.input).map_err(as_transport_error)?,
                codec::packing(codec, cargo.known).map_err(as_transport_error)?,
            ),
        };
        let work = Request::Work {
            plan: plan.clone(),
            input,
            known,
            keys: cargo.keys.to_vec(),
            placement: cargo.placement.clone(),
            memory: cargo.memory.clone(),
        };
        match open.say(&work, seen)? {
            Answer::Done(outcome) => match self.codec.as_deref() {
                None => Ok(outcome),
                Some(codec) => live(codec, outcome),
            },
            Answer::Failed(why) => Err(TransportError::new(why)),
            other => Err(unexpected("working", &other)),
        }
    }
}

/// What came back, alive again.
///
/// A failure here is not a value left behind: everything in this answer was
/// written down by the other side a moment ago, so one that cannot be read back
/// means the two ends do not register the same codecs — and that is the answer,
/// not a value to work around.
fn live(codec: &dyn Codec, outcome: Outcome) -> Result<Outcome, TransportError> {
    Ok(Outcome {
        last: codec.unpacked(&outcome.last).map_err(as_transport_error)?,
        produced: codec::unpacking(codec, &outcome.produced).map_err(as_transport_error)?,
        keys: outcome.keys,
    })
}

fn as_transport_error(e: crate::CodecError) -> TransportError {
    TransportError::new(e.to_string())
}

/// What it answered does not match what it was asked: not a job failure, but
/// the two sides not speaking the same protocol.
fn unexpected(during: &str, answer: &Answer) -> TransportError {
    match answer {
        Answer::Refused(why) => {
            TransportError::new(format!("the worker does not accept the session: {why}"))
        }
        other => TransportError::new(format!(
            "while {during}, the worker answered something beside the point: {other:?}"
        )),
    }
}

/// A pipe failure almost always means the same thing — unless the error
/// already explains itself.
fn broke(e: &std::io::Error) -> TransportError {
    match e.kind() {
        std::io::ErrorKind::InvalidData | std::io::ErrorKind::InvalidInput => {
            TransportError::new(e.to_string())
        }
        _ => TransportError::new(format!(
            "the conversation with the worker was cut off ({e}); usually it has \
             died — check its stderr"
        )),
    }
}

impl Drop for Worker {
    /// Ends the conversation: a child gets its input closed and is waited for;
    /// a standing worker just loses the socket and awaits another client.
    fn drop(&mut self) {
        let mut open = match self.open.lock() {
            Ok(open) => open,
            Err(poisoned) => poisoned.into_inner(),
        };
        if let Link::Child { child, to, .. } = &mut open.link {
            drop(to.take());
            let _ = child.wait();
        }
    }
}
